<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8">
<title>Using an Exception Handler</title>
<style>
table th { border: 1px solid; }
table td { border: 1px solid; }
</style>
</head>
<body>
<hr />
<h2>description: The following examples demonstrate the use of an exception handler.
ms.assetid: c3b4e696-9f45-4616-ac6b-07ba29750bb2
title: Using an Exception Handler
ms.topic: article
ms.date: 05/31/2018</h2>
<h1>Using an Exception Handler</h1>
<p>The following examples demonstrate the use of an exception handler.</p>
<h2>Example 1</h2>
<p>The following code fragment uses structured exception handling to check whether a division operation on two 32-bit integers will result in an division-by-zero error. If this occurs, the function returns <strong>FALSE</strong>â€” otherwise it returns <strong>TRUE</strong>.</p>
<pre lang="C++"><code>BOOL SafeDiv(INT32 dividend, INT32 divisor, INT32 *pResult)
{
    __try 
    { 
        *pResult = dividend / divisor; 
    } 
    __except(GetExceptionCode() == EXCEPTION_INT_DIVIDE_BY_ZERO ? 
             EXCEPTION_EXECUTE_HANDLER : EXCEPTION_CONTINUE_SEARCH)
    { 
        return FALSE;
    }
    return TRUE;
} 
</code></pre>
<h2>Example 2</h2>
<p>The following example function calls the <a href="/windows/win32/api/debugapi/nf-debugapi-debugbreak"><strong>DebugBreak</strong></a> function and uses structured exception handling to check for a breakpoint exception. If one occurs, the function returns <strong>FALSE</strong>â€” otherwise it returns <strong>TRUE</strong>.</p>
<p>The filter expression in the example uses the <a href="getexceptioncode.html"><strong>GetExceptionCode</strong></a> function to check the exception type before executing the handler. This enables the system to continue its search for an appropriate handler if some other type of exception occurs.</p>
<p>Also, use of the <strong>return</strong> statement in the <strong>__try</strong> block of an exception handler differs from the use of <strong>return</strong> in the <strong>__try</strong> block of a termination handler, which causes an abnormal termination of the <strong>__try</strong> block. This is an valid use of the return statement in an exception handler.</p>
<pre lang="C++"><code>BOOL CheckForDebugger()
{
    __try 
    {
        DebugBreak();
    }
    __except(GetExceptionCode() == EXCEPTION_BREAKPOINT ? 
             EXCEPTION_EXECUTE_HANDLER : EXCEPTION_CONTINUE_SEARCH) 
    {
        // No debugger is attached, so return FALSE 
        // and continue.
        return FALSE;
    }
    return TRUE;
}
</code></pre>
<p>Only return EXCEPTION_EXECUTE_HANDLER from an exception filter when the exception type is expected and the faulting address is known. You should allow the default exception handler to process unexpected exception types and faulting addresses.</p>
<h2>Example 3</h2>
<p>The following example shows the interaction of nested handlers. The <a href="/windows/win32/api/errhandlingapi/nf-errhandlingapi-raiseexception"><strong>RaiseException</strong></a> function causes an exception in the guarded body of a termination handler that is inside the guarded body of an exception handler. The exception causes the system to evaluate the FilterFunction function, whose return value in turn causes the exception handler to be invoked. However, before the exception-handler block is executed, the <strong>__finally</strong> block of the termination handler is executed because the flow of control has left the <strong>__try</strong> block of the termination handler.</p>
<pre lang="C++"><code>DWORD FilterFunction() 
{ 
    printf(&quot;1 &quot;);                     // printed first 
    return EXCEPTION_EXECUTE_HANDLER; 
} 
 
VOID main(VOID) 
{ 
    __try 
    { 
        __try 
        { 
            RaiseException( 
                1,                    // exception code 
                0,                    // continuable exception 
                0, NULL);             // no arguments 
        } 
        __finally 
        { 
            printf(&quot;2 &quot;);             // this is printed second 
        } 
    } 
    __except ( FilterFunction() ) 
    { 
        printf(&quot;3\n&quot;);                // this is printed last 
    } 
}
</code></pre>
<p>Â </p>
<p>Â </p>
</body>
